plugins/open_file_env.py

```diff
@@ -0,0 +1,34 @@
+import sublime
+import sublime_plugin
+
+import os
+
+# Related reading
+#     https://forum.sublimetext.com/t/can-i-use-env-variable-in-paths-of-sublime-commands/46437
+
+# The internal open_file command can be used to open specifically named files
+# (say as part of a key binding or menu entry) and supports the $platform and
+# $packages variables, but no others.
+#
+# This plugin implements a drop-in replacement for the open_file command that
+# will not only expand all of the variables usually available in a sublime-
+# build file, but all environment variables as well.
+#
+# The original use case was to share one configuration file across multiple
+# machines where the same files exist in different places by using an
+# environment variable. It could be used for a variety of tasks, however.
+
+class OpenFileEnvCommand(sublime_plugin.WindowCommand):
+    """
+    Drop in replacement for the open_file command that expands environment
+    variables as well as standard Sublime variables in all command arguments.
+    """
+    def run(self, **kwargs):
+        # Create a set of variables based on the current process environment
+        # and the standard Sublime variables
+        variables = dict(os.environ)
+        variables.update(self.window.extract_variables())
+
+        # Expand all variables and execute the base command with them
+        kwargs = sublime.expand_variables(kwargs, variables)
```

Comment:
this confused me until I looked up the API docs and saw that dicts are recursively expanded - nice solution!

Reply:
Yeah I've always found that fairly elegant API solution every time I have cause to expand variables. 

I had a thought that this should maybe expand the `file` argument explicitly to stop it from expanding something that looks like a variable in the `contents` argument (if it's provided). However it seems like the `open_file` command calls `expand_variables()` on the `contents` argument as well, so maybe that's less of an issue.

Regarding the conversation on Discord re: this command requiring extra escaping, that is indeed the case. 

In particular, the underlying `open_file` command seems to call `expand_variables()` with only some variables in the list; so trying to open `/home/tmartin/$bob.txt` actually opens `/home/tmartin/.txt` instead; `$file` is not expanded, but some others are (`$platform`, `$packages`, `$project`, etc).

So in the case of that command, you need to use `/home/tmartin/\\$bob.txt` to open the file. This however does not apply to `window.open_file()`, which does no expansions.

Since we're expanding variables before calling into `open_file` which also expands variables itself, in order to open a file with an internal `$` character you need a path like `/home/tmartin/\\\\\\$bob.txt`. 

I'm not sure if there's a good general solution for something like that (or rather, if it's necessarily worth it); The list of variables that will ultimately get expanded is a known quantity, so in theory perhaps a regex find/replace in the input string could auto-quote any `$` character that doesn't seem to associate with a variable that will expand. 

Perhaps it's worth doing that wholesale for any `$` in the `contents` argument so that it behaves the same as `open_file` would if that file has a `$` in it. 

We may want to merge this one in as-is first and consider these things as an update on it, though. Thoughts?

Reply:
Thanks for the thorough investigation and explanation. I agree that it could make sense to explore it further separately :+1: